server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -101,60 +155,24 @@ public final void write(final T state, final Path... locations) throws IOExcepti
         if (locations.length <= 0) {
             throw new IllegalArgumentException("One or more locations required");
         }
-        final long maxStateId = findMaxStateId(prefix, locations)+1;
+        final long maxStateId = findMaxStateId(prefix, locations) + 1;
         assert maxStateId >= 0 : "maxStateId must be positive but was: [" + maxStateId + "]";
+
         final String fileName = prefix + maxStateId + STATE_FILE_EXTENSION;
-        Path stateLocation = locations[0].resolve(STATE_DIR_NAME);
-        Files.createDirectories(stateLocation);
-        final Path tmpStatePath = stateLocation.resolve(fileName + ".tmp");
-        final Path finalStatePath = stateLocation.resolve(fileName);
-        try {
-            final String resourceDesc = "MetaDataStateFormat.write(path=\"" + tmpStatePath + "\")";
-            try (OutputStreamIndexOutput out =
-                     new OutputStreamIndexOutput(resourceDesc, fileName, Files.newOutputStream(tmpStatePath), BUFFER_SIZE)) {
-                CodecUtil.writeHeader(out, STATE_FILE_CODEC, STATE_FILE_VERSION);
-                out.writeInt(FORMAT.index());
-                try (XContentBuilder builder = newXContentBuilder(FORMAT, new IndexOutputOutputStream(out) {
-                    @Override
-                    public void close() throws IOException {
-                        // this is important since some of the XContentBuilders write bytes on close.
-                        // in order to write the footer we need to prevent closing the actual index input.
-                    } })) {
+        final String tmpFileName = fileName + ".tmp";
+        final Path firstStateLocation = locations[0].resolve(STATE_DIR_NAME);
+        try (Directory stateDir = newDirectory(firstStateLocation)) {
+            writeStateToFirstLocation(state, stateDir, fileName, tmpFileName);
+            logger.trace("written state to {}", firstStateLocation.resolve(fileName));
```

Comment:
It's done in this way because there I did not want to pass firstStateLocation path to writeStateToFirstLocation method just for logging purposes. Unfortunately, there is no method like resolve in Lucene directory. But fixed anyway